ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -873,6 +882,36 @@ protected void maybeProcessCallbacks() {
       }
     }
 
+    /**
+     * Decompress the compressed buffer specified.
+     * If the buffer is not compressed, based on blobCompressed property, then return the input buffer.
+     *
+     * @param sourceBuffer The source buffer to decompress.
+     * @return A new buffer if decompressed successfully and the source buffer is released.
+     *         The source buffer if it is not compressed.
+     *         NULL if decompression failed and the source buffer is released.
+     */
+    protected ByteBuf decompressContent(ByteBuf sourceBuffer) {
+      if (!isChunkCompressed) {
+        // Blob is not compressed.  Return source buffer.
+        return sourceBuffer;
+      }
+
+      try {
+        return decompressionService.decompress(sourceBuffer, routerConfig.routerMaxPutChunkSizeBytes);
+      } catch (Exception ex) {
+        // Note: No need to emit metrics here because decompress() already emits failure metrics.
+        logger.error("Failed to decompress chunk data for chunkBlobId " + chunkBlobId + " due to exception.", ex);
+        setOperationException(buildChunkException("Exception thrown on decompressing the data chunk", ex,
+            RouterErrorCode.UnexpectedInternalError));
+        setOperationCompleted();
```

Comment:
we should just set the operationCompleted=true without calling this method. This method would also release the previous chunk data. We don't have to call it here.

Reply:
In this case, it should release other data chunks right?  If one chunk failed, the entire blob should fail.  Is there a reason why we want to keep the previous chunks of data?